src/sdk/pynni/nni/nas/pytorch/darts/trainer.py

```diff
@@ -0,0 +1,139 @@
+            self.optimizer.zero_grad()
+            self.mutator.reset()
+            logits = self.model(trn_X)
+            loss = self.loss(logits, trn_y)
+            loss.backward()
+            # gradient clipping
+            nn.utils.clip_grad_norm_(self.model.parameters(), 5.)
+            self.optimizer.step()
+
+            new_model = copy.deepcopy(self.model.state_dict())
+
+            # phase 2. architect step (alpha)
+            self.ctrl_optim.zero_grad()
+            # compute unrolled loss
+            self._unrolled_backward(trn_X, trn_y, val_X, val_y, backup_model, lr)
+            self.ctrl_optim.step()
+
+            self.model.load_state_dict(new_model)
+
+            metrics = self.metrics(logits, trn_y)
+            metrics["loss"] = loss.item()
+            meters.update(metrics)
+            if self.log_frequency is not None and step % self.log_frequency == 0:
+                print("Epoch [{}/{}] Step [{}/{}]  {}".format(epoch, self.num_epochs, step, len(self.train_loader), meters))
+
+    def validate_one_epoch(self, epoch):
+        self.model.eval()
+        self.mutator.eval()
+        meters = AverageMeterGroup()
+        with torch.no_grad():
+            self.mutator.reset()
+            for step, (X, y) in enumerate(self.test_loader):
+                X, y = X.to(self.device), y.to(self.device)
+                logits = self.model(X)
+                metrics = self.metrics(logits, y)
+                meters.update(metrics)
+                if self.log_frequency is not None and step % self.log_frequency == 0:
+                    print("Epoch [{}/{}] Step [{}/{}]  {}".format(epoch, self.num_epochs, step, len(self.valid_loader), meters))
+
+    def _unrolled_backward(self, trn_X, trn_y, val_X, val_y, backup_model, lr):
+        """
+        Compute unrolled loss and backward its gradients
+        Parameters
+        ----------
+        v_model: backup model before this step
+        lr: learning rate for virtual gradient step (same as net lr)
+        """
+        self.mutator.reset()
+        loss = self.loss(self.model(val_X), val_y)
+        w_model = tuple(self.model.parameters())
+        w_ctrl = tuple(self.mutator.parameters())
+        w_grads = torch.autograd.grad(loss, w_model + w_ctrl)
+        d_model = w_grads[:len(w_model)]
+        d_ctrl = w_grads[len(w_model):]
+
+        hessian = self._compute_hessian(backup_model, d_model, trn_X, trn_y)
+        with torch.no_grad():
+            for param, d, h in zip(w_ctrl, d_ctrl, hessian):
+                param.grad = d - lr * h
+
+    def _compute_hessian(self, model, dw, trn_X, trn_y):
+        """
+        dw = dw` { L_val(w`, alpha) }
+        w+ = w + eps * dw
+        w- = w - eps * dw
+        hessian = (dalpha { L_trn(w+, alpha) } - dalpha { L_trn(w-, alpha) }) / (2*eps)
+        eps = 0.01 / ||dw||
+        """
+        self.model.load_state_dict(model)
+
+        norm = torch.cat([w.view(-1) for w in dw]).norm()
```

Comment:
does `norm` has any possibility to be `zero`?

Reply:
added a warning.